File: leetcode/binary_search/3073_Maximum_Increasing_Triplet_Value/codes/stack.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maximumTripletValue(vector<int>& nums) {
        int n = nums.size();
        vector<int> max_right(n, 0);
        int right_max = 0;

        for (int k = n - 1; k >= 0; --k) {
            if (nums[k] >= right_max) {
                right_max = nums[k];
            } else {
                max_right[k] = right_max;
            }
        }

        int max_triplet = 0;
        vector<int> idx(n);
        iota(idx.begin(), idx.end(), 0);

        sort(idx.begin(), idx.end(), [&](int a, int b) {
            if (nums[a] == nums[b]) return a > b;
            return nums[a] < nums[b];
        });

        vector<int> stack;
        for (int j : idx) {
            while (!stack.empty() && stack.back() > j) {
                stack.pop_back();
            }
            if (!stack.empty() && max_right[j] >= 0) {
                max_triplet = max(max_triplet, nums[stack.back()] - nums[j] + max_right[j]);
            }
            stack.push_back(j);
        }

        return max_triplet;
    }
};

File: leetcode/binary_search/3073_Maximum_Increasing_Triplet_Value/codes/stack.py
from typing import List

class Solution:
    def maximumTripletValue(self, nums: List[int]) -> int:
        n = len(nums)
        max_right = [0] * n
        right_max = 0

        for k in range(n - 1, -1, -1):
            if nums[k] >= right_max:
                right_max = nums[k]
            else:
                max_right[k] = right_max

        max_triplet = 0
        stack = []

        for j in sorted(range(n), key=lambda x: (nums[x], -x)):
            while stack and stack[-1] > j:
                stack.pop()
            if stack and max_right[j] >= 0:
                max_triplet = max(max_triplet, nums[stack[-1]] - nums[j] + max_right[j])
            stack.append(j)

        return max_triplet
